File: exercise-6-library.kt
data class Book(var title: String, var author: String, var pages: Int, var isBorrowed: Boolean)

class Library {
  private val books: MutableList<Book> = mutableListOf()

  fun addBook(book: Book) {
    books.add(book)
  }

  fun borrowBook(title: String) {
    val book = books.find { it.title == title && !it.isBorrowed }
    if (book != null) {
      book.isBorrowed = true
      println("Livro '${book.title}' emprestado com sucesso!")
    } else {
      println("Livro não disponível para empréstimo.")
    }
  }

  fun returnBorrowedBook(title: String) {
    val book = books.find { it.title == title && it.isBorrowed }
    if (book != null) {
      book.isBorrowed = false
      println("Livro '${book.title}' devolvido com sucesso!")
    } else {
      println("Livro não encontrado ou já disponível.")
    }
  }

  fun isBookAvailable(title: String): Boolean {
    return books.any { it.title == title && !it.isBorrowed }
  }

  fun listAllBooks() {
    println("Coleção de livros na biblioteca:")
    books.forEach {
      println(
              "${it.title} - ${it.author} (${it.pages} páginas) - ${if (it.isBorrowed) "Emprestado" else "Disponível"}"
      )
    }
  }

  fun listBorrowedBooks() {
    val borrowedBooks = books.filter { it.isBorrowed }
    if (borrowedBooks.isEmpty()) {
      println("Nenhum livro emprestado no momento.")
    } else {
      println("Livros emprestados:")
      borrowedBooks.forEach { println("${it.title} - ${it.author}") }
    }
  }

  fun listAvailableBooks() {
    val availableBooks = books.filter { !it.isBorrowed }
    if (availableBooks.isEmpty()) {
      println("Nenhum livro disponível no momento.")
    } else {
      println("Livros disponíveis:")
      availableBooks.forEach { println("${it.title} - ${it.author}") }
    }
  }

  fun getTotalBooks(): Int = books.size

  fun getBorrowedBooksCount(): Int = books.count { it.isBorrowed }

  fun getAvailableBooksCount(): Int = books.count { !it.isBorrowed }
}

fun main() {
  val library = Library()

  val book1 = Book("Livro1", "Autor1", 100, false)
  val book2 = Book("Livro2", "Autor2", 200, true)

  library.addBook(book1)
  library.addBook(book2)

  library.listAllBooks()
  library.listAvailableBooks()

  library.borrowBook("Livro1")
  library.listBorrowedBooks()

  library.returnBorrowedBook("Livro1")
  library.listAvailableBooks()

  println("\nEstatísticas da Biblioteca:")
  println("Total de livros: ${library.getTotalBooks()}")
  println("Livros emprestados: ${library.getBorrowedBooksCount()}")
  println("Livros disponíveis: ${library.getAvailableBooksCount()}")
}

File: exercise-6-library.ts
interface Book {
  title: string;
  author: string;
  pages: number;
  isBorrowed: boolean;
}

class Library {
  private books: Book[] = [];

  addBook(book: Book): void {
    this.books.push(book);
  }

  borrowBook(title: string): void {
    const book = this.books.find((b) => b.title === title && !b.isBorrowed);
    if (!book) {
      console.log("Livro não disponível para empréstimo.");
      return;
    }

    book.isBorrowed = true;
    console.log(`Livro '${book.title}' emprestado com sucesso!`);
  }

  returnBorrowedBook(title: string): void {
    const book = this.books.find((b) => b.title === title && b.isBorrowed);
    if (!book) {
      console.log("Livro não encontrado ou já disponível.");
      return;
    }

    book.isBorrowed = false;
    console.log(`Livro '${book.title}' devolvido com sucesso!`);
  }

  isBookAvailable(title: string): boolean {
    return this.books.some((b) => b.title === title && !b.isBorrowed);
  }

  listAllBooks(): void {
    console.log("Coleção de livros na biblioteca:");
    this.books.forEach((b) => {
      console.log(
        `${b.title} - ${b.author} (${b.pages} páginas) - ${
          b.isBorrowed ? "Emprestado" : "Disponível"
        }`
      );
    });
  }

  listBorrowedBooks(): void {
    const borrowedBooks = this.books.filter((b) => b.isBorrowed);
    if (borrowedBooks.length === 0) {
      console.log("Nenhum livro emprestado no momento.");
      return;
    }

    console.log("Livros emprestados:");
    borrowedBooks.forEach((b) => console.log(`${b.title} - ${b.author}`));
  }

  listAvailableBooks(): void {
    const availableBooks = this.books.filter((b) => !b.isBorrowed);
    if (availableBooks.length === 0) {
      console.log("Nenhum livro disponível no momento.");
      return;
    }

    console.log("Livros disponíveis:");
    availableBooks.forEach((b) => console.log(`${b.title} - ${b.author}`));
  }

  getTotalBooks(): number {
    return this.books.length;
  }

  getBorrowedBooksCount(): number {
    return this.books.filter((b) => b.isBorrowed).length;
  }

  getAvailableBooksCount(): number {
    return this.books.filter((b) => !b.isBorrowed).length;
  }
}

const library = new Library();

const book1: Book = {
  title: "Livro1",
  author: "Autor1",
  pages: 100,
  isBorrowed: false,
};
const book2: Book = {
  title: "Livro2",
  author: "Autor2",
  pages: 200,
  isBorrowed: true,
};

library.addBook(book1);
library.addBook(book2);

library.listAllBooks();
library.listAvailableBooks();

library.borrowBook("Livro1");
library.listBorrowedBooks();

library.returnBorrowedBook("Livro1");
library.listAvailableBooks();

console.log("\nEstatísticas da Biblioteca:");
console.log(`Total de livros: ${library.getTotalBooks()}`);
console.log(`Livros emprestados: ${library.getBorrowedBooksCount()}`);
console.log(`Livros disponíveis: ${library.getAvailableBooksCount()}`);
